util/winutil/restartmgr_windows.go

```diff
@@ -0,0 +1,830 @@
+	if rp != nil {
+		delete(rps, rp.Process.PID)
+	}
+}
+
+// Close invokes (*RestartableProcess).Close on every value in rps, and then
+// clears rps.
+func (rps RestartableProcesses) Close() error {
+	for _, v := range rps {
+		v.Close()
+	}
+	clear(rps)
+	return nil
+}
+
+// _MAXIMUM_WAIT_OBJECTS is the Win32 constant for the maximum number of
+// handles that a call to WaitForMultipleObjects may receive at once.
+const _MAXIMUM_WAIT_OBJECTS = 64
+
+// Terminate forcibly terminates all processes in rps using exitCode, and then
+// waits for their process handles to signal, up to timeout.
+func (rps RestartableProcesses) Terminate(logf logger.Logf, exitCode uint32, timeout time.Duration) error {
+	if len(rps) == 0 {
+		return nil
+	}
+
+	millis, err := wingoes.DurationToTimeoutMilliseconds(timeout)
+	if err != nil {
+		return err
+	}
+
+	errs := []error{}
+	procs := make([]*RestartableProcess, 0, len(rps))
+	handles := make([]windows.Handle, 0, len(rps))
+	for _, v := range rps {
+		if err := windows.TerminateProcess(v.handle, exitCode); err != nil {
+			if err == windows.ERROR_ACCESS_DENIED {
+				// If v terminated before we attempted to terminate, we'll receive
+				// ERROR_ACCESS_DENIED, which is not really an error worth reporting in
+				// our use case. Just obtain the exit code and then close the process.
+				if err := windows.GetExitCodeProcess(v.handle, &v.exitCode); err != nil {
+					logf("GetExitCodeProcess failed: %v", err)
+				} else {
+					v.hasExitCode = true
+				}
+				v.Close()
+			} else {
+				errs = append(errs, &terminationError{rp: v, err: err})
+			}
+			continue
+		}
+		procs = append(procs, v)
+		handles = append(handles, v.handle)
+	}
+
+	for len(handles) > 0 {
+		// WaitForMultipleObjects can only wait on _MAXIMUM_WAIT_OBJECTS handles per
+		// call, so we batch them as necessary.
+		count := uint32(min(len(handles), _MAXIMUM_WAIT_OBJECTS))
+		waitCode, err := windows.WaitForMultipleObjects(handles[:count], true, millis)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("waiting on terminated process handles: %w", err))
+			break
+		}
+		if e := windows.Errno(waitCode); e == windows.WAIT_TIMEOUT {
+			errs = append(errs, fmt.Errorf("waiting on terminated process handles: %w", error(e)))
+			break
+		}
+		if waitCode >= windows.WAIT_OBJECT_0 && waitCode < (windows.WAIT_OBJECT_0+count) {
+			// The first count process handles have all been signaled. Close them out.
+			for _, proc := range procs[:count] {
+				if err := windows.GetExitCodeProcess(proc.handle, &proc.exitCode); err != nil {
+					logf("GetExitCodeProcess failed: %v", err)
+				} else {
+					proc.hasExitCode = true
+				}
+				proc.Close()
+			}
+			procs = procs[count:]
+			handles = handles[count:]
+			continue
+		}
+		// We really shouldn't be reaching this point
+		logf("unexpected state from WaitForMultipleObjects: %d", waitCode)
+	}
+
+	if len(errs) != 0 {
+		return multierr.New(errs...)
+	}
+	return nil
```

Comment:
`clear(rps)` before returning?

Reply:
We don't want to `clear`, actually, because after we have confirmed that we have terminated everything, we still need to serialize that data out to a file (which will happen in a follow-up PR).